clang/lib/Frontend/TextDiagnostic.cpp

```diff
@@ -1349,7 +1349,7 @@ void TextDiagnostic::emitSnippetAndCaret(
   // Prepare source highlighting information for the lines we're about to
   // emit, starting from the first line.
   std::unique_ptr<SmallVector<StyleRange>[]> SourceStyles =
-      highlightLines(BufStart, Lines.first, Lines.second, PP, LangOpts,
+      highlightLines(BufData, Lines.first, Lines.second, PP, LangOpts,
```

Comment:
So this should only matter if there is a null-terminator in the `BufData`, right?  I'm really curious what could have caused that.  It would be appreciated if we could still get a reduced version of the test for a lit-test, but as this is blocking your downstream, I can accept without a repro.

Reply:
I've got a (still quite hairy and not shareable) reproducer that requires the use of a number of defines on the command line to actually cause an assertion failure (`-D_Q1 -D_Q2 -D_Q3 -D_Q4 -D_Q5 -D_Q6 -D_Q7 -D_Q8 -D_Q9 -D_Q10 -D_Q11 -D_Q12 -D_Q13  -D_Q14  -D_Q15 -D_Q16 -D_Q17 -D_Q18 -D_Q19`). Removing any of the defines makes the crash disappear.

Reply:
This may explain the presence of \0's

Reply:
I'll try to provide what I have so far, but I doubt it's feasible to create a good test case for this.

So here's what I have so far (requires GLib 2.78.3 in third_party/glib/):
```
$ cat q.c
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include "q.h"
$ cat q.h
#ifndef __Q_H__
#define __Q_H__

#include <stdio.h>
#include "third_party/glib/glib/glib.h"

G_INLINE_FUNC void q();

#endif
$ cat stl/math.h
#ifndef _STL_MATH_H_
#define _STL_MATH_H_

#include_next <math.h>

#endif
$ cat stl/string.h
#ifndef _STL_STRING_H_
#define _STL_STRING_H_

#include_next <string.h>

#endif
$ clang -iquote . -isystem stl  -fdiagnostics-show-option -fcolor-diagnostics  -Wall -Werror --target=aarch64-unknown-linux-gnu -mcpu=neoverse-n1  --sysroot=our/custom/libc -D_Q1 -D_Q2 -D_Q3 -D_Q4 -D_Q5 -D_Q6 -D_Q7 -D_Q8 -D_Q9 -D_Q10 -D_Q11 -D_Q12 -D_Q13  -D_Q14  -D_Q15 -D_Q16 -D_Q17 -D_Q18 -D_Q19 -c q.c -fsyntax-only

In file included from q.c:4:
./q.h:7:1: warning: Deprecated pre-processor symbol: replace with "static inline" [-W#pragma-messages]
    7 | G_INLINE_FUNC void q();
      | ^
./third_party/glib/glib/gmacros.h:157:39: note: expanded from macro 'G_INLINE_FUNC'
  157 | #  define G_INLINE_FUNC static inline GLIB_DEPRECATED_MACRO_IN_2_48_FOR(static inline)
      |                                       ^
./third_party/glib/glib/glib-visibility.h:414:46: note: expanded from macro 'GLIB_DEPRECATED_MACRO_IN_2_48_FOR'
  414 | #define GLIB_DEPRECATED_MACRO_IN_2_48_FOR(f) GLIB_DEPRECATED_MACRO_FOR (f)
      |                                              ^
./third_party/glib/glib/gmacros.h:1299:3: note: expanded from macro 'GLIB_DEPRECATED_MACRO_FOR'
 1299 |   _GLIB_GNUC_DO_PRAGMA(GCC warning G_STRINGIFY (Deprecated pre-processor symbol: replace with #f))
      |   ^
./third_party/glib/glib/gmacros.h:1296:33: note: expanded from macro '_GLIB_GNUC_DO_PRAGMA'
 1296 | #define _GLIB_GNUC_DO_PRAGMA(x) _Pragma(G_STRINGIFY (x))                                                                                                                                                                                           |                                 ^                                                                                                                                                                                                            <scratch space>:40:6: note: expanded from here
assert.h assertion failed at llvm-project/clang/lib/Frontend/TextDiagnostic.cpp:1157 in std::unique_ptr<llvm::SmallVector<TextDiagnostic::StyleRange>[]> highlightLines(StringR
ef, unsigned int, unsigned int, const Preprocessor *, const LangOptions &, bool, FileID, const SourceManager &): CheckPoint >= Buff->getBufferStart() && CheckPoint <= Buff->getBufferEnd()
*** Check failure stack trace: ***
    @     0x562b1f8b03a3  __assert_fail
    @     0x562b170c558e  clang::TextDiagnostic::emitSnippetAndCaret()
    @     0x562b170c7b02  clang::TextDiagnostic::emitCodeContext()
    @     0x562b1709bcbc  clang::DiagnosticRenderer::emitDiagnostic()
    @     0x562b1709ef73  clang::DiagnosticRenderer::emitSingleMacroExpansion()
    @     0x562b1709cf56  clang::DiagnosticRenderer::emitMacroExpansions()
    @     0x562b1709bd47  clang::DiagnosticRenderer::emitDiagnostic()
    @     0x562b170bc4bb  clang::TextDiagnosticPrinter::HandleDiagnostic()
    @     0x562b1996038a  clang::DiagnosticIDs::EmitDiag()
    @     0x562b19960183  clang::DiagnosticIDs::ProcessDiag()
    @     0x562b199564ac  clang::DiagnosticsEngine::EmitCurrentDiagnostic()
    @     0x562b198aa621  (anonymous namespace)::PragmaMessageHandler::HandlePragma()
    @     0x562b198a12d6  clang::PragmaNamespace::HandlePragma()
    @     0x562b198a12d6  clang::PragmaNamespace::HandlePragma()
    @     0x562b198a1587  clang::Preprocessor::HandlePragmaDirective()
    @     0x562b198a21ad  clang::Preprocessor::Handle_Pragma()
    @     0x562b1988aec3  clang::Preprocessor::ExpandBuiltinMacro()
    @     0x562b19889599  clang::Preprocessor::HandleMacroExpandedIdentifier()
    @     0x562b198c32f8  clang::Preprocessor::HandleIdentifier()
    @     0x562b198d5023  clang::TokenLexer::Lex()
    @     0x562b1987c316  clang::Preprocessor::CLK_TokenLexer()
    @     0x562b198c3bfd  clang::Preprocessor::Lex()
    @     0x562b173942fa  clang::Parser::ConsumeToken()
    @     0x562b17393915  clang::Parser::ConsumeAnyToken()
    @     0x562b174784f5  clang::Parser::ParseDeclarationSpecifiers()
    @     0x562b1739ed03  clang::Parser::ParseDeclOrFunctionDefInternal()
    @     0x562b1739e4af  clang::Parser::ParseDeclarationOrFunctionDefinition()
    @     0x562b1739c864  clang::Parser::ParseExternalDeclaration()                                                                                                                                                                                      @     0x562b17399164  clang::Parser::ParseTopLevelDecl()
```

The problem is that I can't cut a reasonable subset of our libc without making the crash disappear.

I hope that's enough to understand the problem.

Reply:
> we have an existing test that embeds a null character into the source and that doesn't seem to reproduce the issue

Can you construct a test case with an embedded null character into the source and make the call to `PP->getCheckPoint(FID, FirstLineStart))` (at `clang/lib/Frontend/TextDiagnostic.cpp:1155` also shown below) return an address _after_ the null? That would reproduce the issue.

```
  const char *FirstLineStart =
      FileData.data() +
      SM.getDecomposedLoc(SM.translateLineCol(FID, StartLineNumber, 1)).second;
  if (const char *CheckPoint = PP->getCheckPoint(FID, FirstLineStart)) {
    assert(CheckPoint >= Buff->getBufferStart() &&
           CheckPoint <= Buff->getBufferEnd());
    assert(CheckPoint <= FirstLineStart);
    size_t Offset = CheckPoint - Buff->getBufferStart();
    L.seek(Offset, /*IsAtStartOfLine=*/false);
  }
```

Reply:
Some debugging.

Backtrace:
```
    frame #7: 0x00005555754b2ee6 clang`__assert_fail(assertion="CheckPoint >= Buff->getBufferStart() && CheckPoint <= Buff->getBufferEnd()", file="llvm-project/clang/lib/Frontend/TextDiagnostic.cpp", line=1157, function="std::unique_ptr<llvm::SmallVector<TextDiagnostic::StyleRange>[]> highlightLines(StringRef, unsigned int, unsigned int, const Preprocessor *, const LangOptions &, bool, FileID, const SourceManager &)") at logging.cc:57:3
    frame #8: 0x0000555567bfc020 clang`highlightLines(FileData=(Data = "\n clang diagnostic ignored \"-Wdeprecated-declarations\"\n", Length = 55), StartLineNumber=40, EndLineNumber=40, PP=0x000025c8bf806018, LangOpts=0x000025c8bf848918, ShowColors=true, FID=(ID = 23810), SM=0x000025c8bf892000) at TextDiagnostic.cpp:1156:5
    frame #9: 0x0000555567bfabcc clang`clang::TextDiagnostic::emitSnippetAndCaret(this=0x000025c8bfc62f00, Loc=FullSourceLoc @ 0x00007fffffff3b68, Level=Note, Ranges=0x00007fffffff3c80, Hints=ArrayRef @ 0x00007fffffff3b90) at TextDiagnostic.cpp:1352:7
    frame #10: 0x0000555567c021a7 clang`clang::TextDiagnostic::emitCodeContext(this=0x000025c8bfc62f00, Loc=FullSourceLoc @ 0x00007fffffff3c00, Level=Note, Ranges=0x00007fffffff3c80, Hints=ArrayRef @ 0x00007fffffff3c20) at TextDiagnostic.h:97:5
    frame #11: 0x0000555567bc1a9d clang`clang::DiagnosticRenderer::emitCaret(this=0x000025c8bfc62f00, Loc=FullSourceLoc @ 0x00007fffffff3ce0, Level=Note, Ranges=ArrayRef @ 0x00007fffffff3cd0, Hints=ArrayRef @ 0x00007fffffff3d00) at DiagnosticRenderer.cpp:429:3
    frame #12: 0x0000555567bc1112 clang`clang::DiagnosticRenderer::emitDiagnostic(this=0x000025c8bfc62f00, Loc=FullSourceLoc @ 0x00007fffffff4270, Level=Note, Message=(Data = "expanded from hereo '_GLIB_GNUC_DO_PRAGMA'_FOR'_48_FOR'", Length = 18), Ranges=ArrayRef @ 0x00007fffffff42a0, FixItHints=ArrayRef @ 0x00007fffffff42b0, D=clang::DiagOrStoredDiag @ 0x00007fffffff4258) at DiagnosticRenderer.cpp:127:5
    frame #13: 0x0000555567bc3549 clang`clang::DiagnosticRenderer::emitSingleMacroExpansion(this=0x000025c8bfc62f00, Loc=FullSourceLoc @ 0x00007fffffff44c8, Level=Warning, Ranges=ArrayRef @ 0x00007fffffff44b8) at DiagnosticRenderer.cpp:454:3
    frame #14: 0x0000555567bc1f4b clang`clang::DiagnosticRenderer::emitMacroExpansions(this=0x000025c8bfc62f00, Loc=FullSourceLoc @ 0x00007fffffff4840, Level=Warning, Ranges=ArrayRef @ 0x00007fffffff4830, Hints=ArrayRef @ 0x00007fffffff4860) at DiagnosticRenderer.cpp:569:7
    frame #15: 0x0000555567bc11dc clang`clang::DiagnosticRenderer::emitDiagnostic(this=0x000025c8bfc62f00, Loc=FullSourceLoc @ 0x00007fffffff4dd0, Level=Warning, Message=(Data = "Deprecated pre-processor symbol: replace with \"static inline\" [-W#pragma-messages]\xff\xff\xff\U0000007f", Length = 82), Ranges=ArrayRef @ 0x00007fffffff4e00, FixItHints=ArrayRef @ 0x00007fffffff4e10, D=clang::DiagOrStoredDiag @ 0x00007fffffff4db8) at DiagnosticRenderer.cpp:132:7
    frame #16: 0x0000555567bf73b6 clang`clang::TextDiagnosticPrinter::HandleDiagnostic(this=0x000025c8bfc16630, Level=Warning, Info=0x00007fffffff50c8) at TextDiagnosticPrinter.cpp:151:13
    frame #17: 0x000055556c6a0148 clang`clang::DiagnosticIDs::EmitDiag(this=0x000025c8bfc04840, Diag=0x000025c8bf85c900, DiagLevel=Warning) const at DiagnosticIDs.cpp:823:16
    frame #18: 0x000055556c69fef0 clang`clang::DiagnosticIDs::ProcessDiag(this=0x000025c8bfc04840, Diag=0x000025c8bf85c900) const at DiagnosticIDs.cpp:815:3
    frame #19: 0x000055556c69437a clang`clang::DiagnosticsEngine::ProcessDiag(this=0x000025c8bf85c900) at Diagnostic.h:1042:19
    frame #20: 0x000055556c68ef05 clang`clang::DiagnosticsEngine::EmitCurrentDiagnostic(this=0x000025c8bf85c900, Force=false) at Diagnostic.cpp:545:15
    frame #21: 0x0000555562f61b4d clang`clang::DiagnosticBuilder::Emit(this=0x00007fffffff5420) at Diagnostic.h:1325:28
    frame #22: 0x0000555562f5fb6c clang`clang::DiagnosticBuilder::~DiagnosticBuilder(this=0x00007fffffff5420) at Diagnostic.h:1366:26
    frame #23: 0x000055556c5db9c0 clang`(anonymous namespace)::PragmaMessageHandler::HandlePragma(this=0x000025c8bfc62c40, PP=0x000025c8bf806018, Introducer=PragmaIntroducer @ 0x00007fffffff54f8, Tok=0x00007fffffff5720) at Pragma.cpp:1675:5
    frame #24: 0x000055556c5d33fb clang`clang::PragmaNamespace::HandlePragma(this=0x000025c8bfc636c0, PP=0x000025c8bf806018, Introducer=PragmaIntroducer @ 0x00007fffffff55c8, Tok=0x00007fffffff5720) at Pragma.cpp:124:12
    frame #25: 0x000055556c5d33fb clang`clang::PragmaNamespace::HandlePragma(this=0x000025c8bfc63600, PP=0x000025c8bf806018, Introducer=PragmaIntroducer @ 0x00007fffffff5698, Tok=0x00007fffffff5720) at Pragma.cpp:124:12
    frame #26: 0x000055556c5d3557 clang`clang::Preprocessor::HandlePragmaDirective(this=0x000025c8bf806018, Introducer=PragmaIntroducer @ 0x00007fffffff5748) at Pragma.cpp:178:19
    frame #27: 0x000055556c5d3fdf clang`clang::Preprocessor::Handle_Pragma(this=0x000025c8bf806018, Tok=0x000025c8bf9c2010) at Pragma.cpp:300:3
    frame #28: 0x000055556c5b1892 clang`clang::Preprocessor::ExpandBuiltinMacro(this=0x000025c8bf806018, Tok=0x000025c8bf9c2010) at PPMacroExpansion.cpp:1506:12
    frame #29: 0x000055556c5b0af3 clang`clang::Preprocessor::HandleMacroExpandedIdentifier(this=0x000025c8bf806018, Identifier=0x000025c8bf9c2010, M=0x00007fffffff6da0) at PPMacroExpansion.cpp:493:5
    frame #30: 0x000055556c5fe271 clang`clang::Preprocessor::HandleIdentifier(this=0x000025c8bf806018, Identifier=0x000025c8bf9c2010) at Preprocessor.cpp:815:18
    frame #31: 0x000055556c6169f7 clang`clang::TokenLexer::Lex(this=0x000025c8bfc16a20, Tok=0x000025c8bf9c2010) at TokenLexer.cpp:717:17
    frame #32: 0x000055556c5a55dd clang`clang::Preprocessor::CLK_TokenLexer(P=0x000025c8bf806018, Result=0x000025c8bf9c2010) at Preprocessor.h:2926:29
    frame #33: 0x000055556c5fe790 clang`clang::Preprocessor::Lex(this=0x000025c8bf806018, Result=0x000025c8bf9c2010) at Preprocessor.cpp:872:11
    frame #34: 0x0000555567eccf2b clang`clang::Parser::ConsumeToken(this=0x000025c8bf9c2000) at Parser.h:518:8
    frame #35: 0x0000555567eccb54 clang`clang::Parser::ConsumeAnyToken(this=0x000025c8bf9c2000, ConsumeCodeCompletionTok=false) at Parser.h:556:12
    frame #36: 0x0000555567fc1d93 clang`clang::Parser::ParseDeclarationSpecifiers(this=0x000025c8bf9c2000, DS=0x00007fffffff9188, TemplateInfo=0x00007fffffff90c0, AS=AS_none, DSContext=DSC_top_level, LateAttrs=0x0000000000000000, AllowImplicitTypename=Yes) at ParseDecl.cpp:4667:7
```

```
frame #8: 0x0000555567bfc020 clang`highlightLines(FileData=(Data = "\n clang diagnostic ignored \"-Wdeprecated-declarations\"\n", Length = 55), StartLineNumber=40, EndLineNumber=40, PP=0x000025c8bf806018, LangOpts=0x000025c8bf848918, ShowColors=true, FID=(ID = 23810), SM=0x000025c8bf892000) at TextDiagnostic.cpp:1156:5
   1153       FileData.data() +
   1154       SM.getDecomposedLoc(SM.translateLineCol(FID, StartLineNumber, 1)).second;
   1155   if (const char *CheckPoint = PP->getCheckPoint(FID, FirstLineStart)) {
-> 1156     assert(CheckPoint >= Buff->getBufferStart() &&
   1157            CheckPoint <= Buff->getBufferEnd());
   1158     assert(CheckPoint <= FirstLineStart);
   1159     size_t Offset = CheckPoint - Buff->getBufferStart();
(lldb) p Buff->BufferEnd
(const char *) 0x000025c8bf052567 ""
(lldb) p CheckPoint
(const char *) 0x000025c8bf05284a "\n"
(lldb) x/1024b Buff->BufferStart
0x25c8bf052530: \n clang diagnostic ignored "-Wde
0x25c8bf052550: precated-declarations"\n\0\nglib_au
0x25c8bf052570: toptr_clear_GUri\0\nglib_autoptr_c
0x25c8bf052590: leanup_GUri\0\nglib_autoptr_clear_
0x25c8bf0525b0: GUri\0\nglib_listautoptr_cleanup_G
0x25c8bf0525d0: Uri\0\nglib_slistautoptr_cleanup_G
0x25c8bf0525f0: Uri\0\nglib_queueautoptr_cleanup_G
0x25c8bf052610: Uri\0\n clang diagnostic pop\n\0\nGPa
0x25c8bf052630: thBuf_autoptr\0\nGPathBuf_listauto
0x25c8bf052650: ptr\0\nGPathBuf_slistautoptr\0\nGPat
0x25c8bf052670: hBuf_queueautoptr\0\n clang diagno
0x25c8bf052690: stic push\n\0\n clang diagnostic ig
0x25c8bf0526b0: nored "-Wdeprecated-declarations
0x25c8bf0526d0: "\n\0\nglib_autoptr_clear_GPathBuf\0
0x25c8bf0526f0: \nglib_autoptr_cleanup_GPathBuf\0\n
0x25c8bf052710: glib_autoptr_clear_GPathBuf\0\ngli
0x25c8bf052730: b_listautoptr_cleanup_GPathBuf\0\n
0x25c8bf052750: glib_slistautoptr_cleanup_GPathB
0x25c8bf052770: uf\0\nglib_queueautoptr_cleanup_GP
0x25c8bf052790: athBuf\0\n clang diagnostic pop\n\0\n
0x25c8bf0527b0:  clang diagnostic push\n\0\n clang
0x25c8bf0527d0: diagnostic ignored "-Wdeprecated
0x25c8bf0527f0: -declarations"\n\0\nglib_auto_clean
0x25c8bf052810: up_GPathBuf\0\n clang diagnostic p
0x25c8bf052830: op\n\0\n clang diagnostic pop\n\0\n"st
0x25c8bf052850: atic inline"\0\n"Deprecated pre-pr
0x25c8bf052870: ocessor symbol: replace with \"s
0x25c8bf052890: tatic inline\""\0\n"GCC warning \"
0x25c8bf0528b0: Deprecated pre-processor symbol:
0x25c8bf0528d0:  replace with \\\"static inline\
0x25c8bf0528f0: \\"\""\0\n GCC warning "Deprecated
0x25c8bf052910:  pre-processor symbol: replace w
```

So the buffer is inside the scratch space, which naturally contains \0's. And the problem is when the code to highlight is anywhere beyond the first \0 character in this buffer - then CheckPoint will be outside of the StringRef created from just the pointer to the buffer start.

Reply:
That's about as much as we can help with further debugging. As I said, I don't see how to create a reliable test case here.